### ISSUE TITLE
[jit] cannot trace tensor factory methods

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#9069]
Tracing newly created tensors is not too much a problem..... until we hit random numbers.

Things like `torch.randn(3)` are traced as `prim::Constant`, e.g.,
```py
>>> def fn(x): return torch.randn(3,3)+x
...
>>>
>>> torch.jit.get_trace_graph(fn, torch.ones(1), nderivs=0)[0].graph()
graph(%0 : Float(1)) {
  %1 : Float(3, 3!) = aten::expand[size=[3, 3], implicit=1](%0)
  %2 : Float(3, 3) = prim::Constant[value= 1.0159 -0.3672 -0.7744  2.4991 -0.6615 -1.9963 -0.1653 -0.6848 -0.5436 [ CPUFloatTensor{3,3} ]]()
  %3 : Float(3, 3) = aten::add[alpha={1}](%2, %1)
  return (%3);
}
```

The `_like` variant is traced similarly (requires_grad or not):
```py
>>> def fn(x): return torch.randn_like(x,requires_grad=True)+x
...
>>> torch.jit.get_trace_graph(fn, torch.ones(1), nderivs=0)[0].graph()
graph(%0 : Float(1)) {
  %1 : Float(1) = prim::Constant[value={0.299761}]()
  %2 : Float(1) = aten::add[alpha={1}](%1, %0)
  return (%2);
}
```

In-place sampling methods like `normal_` aren't supported either.

This is blocking moving dropout into ATen #9008 . Because, assuming we don't have custom backward for dropout (we really shouldn't since it's just sampling and multiplication), then we need to trace creation of the stochastic mask tensor.

The workaround that I really really want to avoid is to have a custom backward. Then, because we need the stochastic mask in the backward, we need to write a `dropout_with_mask` that returns both the output and the mask, and write `dropout` as a wrapper. This shouldn't be necessary for such a simple op and makes the mask depending on input in graph.

The fundamental solution is to fix the jit tracer on random numbers.

Relevant #8450 .

cc @apaszke  @jamesr66a

**Comment by apaszke** [comment]
This is kind of expected. There are two main problems at play here:

1. RNG calls are not functional - they mutate the state of the global PRNG, and therefore we probably shouldn't rearrange them. This however requires us to add the "world token" to inhibit optimizations.
2. We never trace constructors. If we want to do this, we probably need to make the tracing a thread-local property, instead of auto-detecting it as a per-Variable property.

**Comment by apaszke** [comment]
Speaking of custom backward, this should be a non-issue. We'll need to add the random ops as non-differentiable in the symbolic AD, but capturing of the mask should happen automatically as part of the derivative for mul.

**Comment by SsnL** [comment]
>>> Speaking of custom backward, this should be a non-issue. We'll need to add the random ops as non-differentiable in the symbolic AD, but capturing of the mask should happen automatically as part of the derivative for mul.

@apaszke I am not very familiar with the jit structure, and not sure that I'm understanding this. Could you elaborate? 

About dropout, my concern was that currently dropout is traced a single `Dropout` op, hiding the random sampling step inside. So the traced graph can be run "as-is" with proper dropout mask sampled very time (from my understanding). But since random ops (`bernoulli` in this case) is traced as `prim::Constant`, then the graph traced is just wrong.

My understanding is probably wrong somewhere. Please correct me if so!

**Comment by SsnL** [comment]
I guess my question was how custom backward and symbolic AD are connected? My understanding was that jit also traces backward, but it seems that that was wrong and backward is handled by symbolic AD?

**Comment by apaszke** [comment]
It's true that we used to trace backward, but it's no longer the case. Right now we separate the graph into parts we can differentiate symbolically, and other ops that should be taken care of by autograd (and we don't have visibility into their backward graph anymore).

**Comment by jamesr66a** [comment]
Closed by https://github.com/pytorch/pytorch/pull/10935 and https://github.com/pytorch/pytorch/pull/11539
